File: benchmarks/fixtures/path_traversal.js
const path = require("path");

// Read-only benchmark fixture: these functions are parsed, never called.
function extractVulnerable(output, entry, writeFile) {
  const destination = path.join(output, entry.path);
  return writeFile(destination, entry.data);
}

function extractPatched(output, entry, writeFile) {
  const destination = path.resolve(output, entry.path);
  const resolvedOutput = path.resolve(output) + path.sep;
  if (!destination.startsWith(resolvedOutput)) {
    throw new Error("archive entry escapes destination");
  }
  return writeFile(destination, entry.data);
}

module.exports = { extractVulnerable, extractPatched };

File: benchmarks/fixtures/path_traversal.py
"""Read-only benchmark fixture. The functions are parsed and never called."""

import os
from pathlib import Path


def extract_vulnerable(root, filename, write_file):
    destination = os.path.join(root, filename)
    return write_file(destination)


def extract_patched(root, filename, write_file):
    destination = Path(root).joinpath(filename)
    destination.relative_to(Path(root))
    return write_file(destination)
